src/toxcore/dht_new/dht_impl.rs

```diff
@@ -0,0 +1,259 @@
+            },
+        };
+
+        match self.packet_kind {
+            PacketKind::PingReq => {
+                if let IResult::Done(_, packet) = PingRequest::from_bytes(&decrypted) {
+                    Ok(Some(DhtPacketPayload::PingRequest(packet)))                    
+                } else {
+                    Ok(None)
+                }
+            },
+            PacketKind::PingResp => {
+                if let IResult::Done(_, packet) = PingResponse::from_bytes(&decrypted) {
+                    Ok(Some(DhtPacketPayload::PingResponse(packet)))
+                } else {
+                    Ok(None)
+                }
+            },
+            PacketKind::GetN => {
+                if let IResult::Done(_, packet) = GetNodes::from_bytes(&decrypted) {
+                    Ok(Some(DhtPacketPayload::GetNodes(packet)))
+                } else {
+                    Ok(None)
+                }
+            },
+            PacketKind::SendN => {
+                if let IResult::Done(_, packet) = SendNodes::from_bytes(&decrypted) {
+                    Ok(Some(DhtPacketPayload::SendNodes(packet)))
+                } else {
+                    Ok(None)
+                }
+            },
+            _ => Ok(None)
+        }
+    }
+
+    /**
+    Create DHT Packet with [`Ping`](./struct.Ping.html) response to `Ping`
+    request that packet contained.
+
+    Nonce for the response is automatically generated.
+    */
+    // Because UDP codec and tokio use DhtBase for send/receive packet, 
+    // this function returns DhtBase type object
+    pub fn ping_response(&self,
+                     secret_key: &SecretKey,
+                     symmetric_key: &PrecomputedKey,
+                     own_public_key: &PublicKey) -> Option<DhtBase> {
+
+        debug!(target: "DhtPacket", "Creating Ping response from Ping request that DHT packet contained.");
+        trace!(target: "DhtPacket", "With args: DhtPacket: {:?}, own_pk: {:?}", self, own_public_key);
+
+        if self.packet_kind != PacketKind::PingReq {
+            return None
+        }
+
+        if let Ok(Some(DhtPacketPayload::PingRequest(packet))) = self.get_payload(secret_key) {
+            let resp = DhtPacketPayload::PingResponse(PingResponse::from(packet));
+            Some(DhtBase::DhtPacket(DhtPacket::new(symmetric_key, own_public_key, resp)))
+        }
+        else {
+            None
+        }
+    }
+}
+
+impl DhtRequest {
+    /// create new DhtRequest object
+    pub fn new(shared_secret: &PrecomputedKey, rpk: &PublicKey, spk: &PublicKey, dp: DhtRequestPayload) -> DhtRequest {
+        let nonce = &gen_nonce();
+
+        let mut buf = [0; MAX_DHT_PACKET_SIZE];
+        let payload = seal_precomputed(dp.to_bytes((&mut buf, 0)).ok().unwrap().0, nonce, shared_secret);
+
+        DhtRequest {
+            rpk: *rpk,
+            spk: *spk,
+            nonce: *nonce,
+            payload: payload,
+        }
+    }
+}
+
+impl GetNodes {
```

Comment:
I was reading `DhtRequest` and then appeared `GetNodes`. It is unexpected

Reply:
Do you mean that the appearing order is not good?
Or impl GetNodes{...} should not be here?

Reply:
Well, you may either move it to change the order or move it to another file.

Reply:
Now, dht_impl.rs has collection of impl of Types.
This concept should be discussed more.